File: src/main/java/renderer/PixelBufferObject.java
package renderer;

import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.opengl.GL15.*;
import static org.lwjgl.opengl.GL21.*;

import java.nio.ByteBuffer;

import engine.AppContext;
import org.lwjgl.BufferUtils;
import renderer.constants.GlTextureTarget;

public class PixelBufferObject {
	
	private int id;
	private int width;
	private int height;
	private ByteBuffer buffer;
	private float[] array;

	public PixelBufferObject(int width, int height) {
		id = glGenBuffers();
		this.width = width;
		this.height = height;
		this.buffer = BufferUtils.createByteBuffer(4*4*width*height); // 4 is byte size of float
		this.array = new float[4*width*height];
		bind();
		glBufferData(GL_PIXEL_UNPACK_BUFFER, 4*4*width*height, GL_STREAM_READ);
		unbind();
	}

	private void unbind() {
		glBindBuffer(GL_PIXEL_UNPACK_BUFFER, 0);
	}
	
	public void bind() {
		glBindBuffer(GL_PIXEL_UNPACK_BUFFER, id);
	}
	
	public void readPixelsFromTexture(int textureId, int mipmapLevel, GlTextureTarget target, int format, int type) {
		bind();
		AppContext.getInstance().getRenderer().getOpenGLContext().bindTexture(target, textureId);
		glGetTexImage(target.glTarget, mipmapLevel, format, type, buffer);
		unbind();
	}
	
	public float[] mapBuffer() {
		glMapBuffer(GL_PIXEL_UNPACK_BUFFER, GL_READ_WRITE, buffer);
		buffer.rewind();
		buffer.asFloatBuffer().get(array);
		glUnmapBuffer(GL_PIXEL_PACK_BUFFER);
		return array;
	}
}

File: opengl/src/main/kotlin/de/hanno/hpengine/graphics/texture/PixelBufferObject.kt
package de.hanno.hpengine.graphics.texture


import de.hanno.hpengine.graphics.GraphicsApi
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL15
import org.lwjgl.opengl.GL21.*
import java.nio.ByteBuffer

context(GraphicsApi)
class PixelBufferObject(val id: Int) {

    companion object {
        context(GraphicsApi)
        operator fun invoke() = PixelBufferObject(onGpu { glGenBuffers() })
    }

    fun put(data: ByteBuffer) {
        val buffer = onGpu {
            GL15.glBindBuffer(GL_PIXEL_UNPACK_BUFFER, id)
            GL15.glBufferData(GL_PIXEL_UNPACK_BUFFER, data.capacity().toLong(), GL15.GL_STREAM_COPY)
            val buffer = GL15.glMapBuffer(GL_PIXEL_UNPACK_BUFFER, GL15.GL_WRITE_ONLY, null)!!
            buffer.rewind()
            buffer
        }

        buffer.put(data)
        unmap()
    }

    fun unmap() = onGpu {
        bind()
        GL15.glUnmapBuffer(GL_PIXEL_UNPACK_BUFFER)
    }

    fun bind() = onGpu {
        GL15.glBindBuffer(GL_PIXEL_UNPACK_BUFFER, id)
    }
    fun unbind() = onGpu {
        GL15.glBindBuffer(GL_PIXEL_UNPACK_BUFFER, 0)
    }
}